### PR TITLE
Use unmanaged in domain maps and in reductions

### LINKED ISSUE #9238: delete-free subtask: convert domain map hierarchy to use unmanaged pointers
As a Chapel implementer, I need the modules and tests to pass as progress continues on the delete-free implementation. To enable that progress, I need the modules and tests to use raw, shared, or owned pointers so that the lifetime checker can work correctly.

This issue is specifically about converting the domain map hierarchy to use unmanaged pointers. That includes _array, _domain, _distribution, as well as the classes providing the implementations for these (e.g. BlockDist, BlockDom, BlockArr...)

Acceptance Criteria:
A version of the compiler with this change passes testing.

### PR DESCRIPTION
**Pull request by mppf**
This PR adjusts the following types supported by the compiler and standard/internal modules to use unmanaged class types:
 * reductions (user defined or included in the standard modules)
 * domain, array, and distribution implementation classes
 * domain, array, and distribution wrapper records

Changing these types to unmanaged allows us to work towards changing the meaning of class types and `new` on classes.

This PR includes several workarounds:
 * _to_unmanaged(this) is a workaround for issue #9274
 * _to_borrowed(sometype) in where clauses is a workaround for issue #9273

Closes #9238.

- [x] full local testing & gasnet testing

Reviewed by @benharsh, @vasslitvinov, @psahabu, and @ben-albrecht - thanks!

### CODE REVIEW

**Comment by benharsh** [comment]
Not that it needs to be done in this PR, but we should think about how to document this change in the DSI technote.

File: modules/dists/PrivateDist.chpl [hunk 1]
```diff
@@ -203,5 +205,5 @@ proc PrivateArr.dsiSerialWrite(x) {
 }
 
 // TODO: Fix 'new Private()' leak -- Discussed in #6726
-const PrivateSpace: domain(1) dmapped new dmap(new Private());
+const PrivateSpace: domain(1) dmapped Private();
```
Comment (benharsh) [inline modules/dists/PrivateDist.chpl]: Does this comment or the related issue need updating?

File: modules/internal/ChapelBase.chpl [hunk 2]
```diff
@@ -1278,6 +1278,15 @@ module ChapelBase {
     __primitive("call destructor", x);
   }
 
+  /*
+  inline proc chpl__tounmanaged(ref arg:Owned) {
+    return arg.release();
+  }
+  inline proc chpl__tounmanaged(arg) where arg:object {
+    return arg;
+  }*/
```
Comment (mppf) [inline modules/internal/ChapelBase.chpl]: Need to delete these

File: modules/internal/ChapelError.chpl [hunk 3]
```diff
@@ -338,7 +338,7 @@ module ChapelError {
   proc chpl_fix_thrown_error(err: Error): Error {
     var fixErr: Error = err;
     if fixErr == nil then
-      fixErr = new NilThrownError();
+      fixErr = new unmanaged NilThrownError();
```
Comment (mppf) [inline modules/internal/ChapelError.chpl]: @psahabu - could you review the changes here to ChapelError? thanks.

File: modules/standard/SysError.chpl [hunk 4]
```diff
@@ -75,42 +75,42 @@ class SystemError : Error {
   */
   proc type fromSyserr(err: syserr, details: string = "") {
     if err == EAGAIN || err == EALREADY || err == EWOULDBLOCK || err == EINPROGRESS {
-      return new BlockingIOError(details, err);
+      return new unmanaged BlockingIOError(details, err);
```
Comment (mppf) [inline modules/standard/SysError.chpl]: @psahabu - could you review changes in this file too? thanks

**Comment by mppf** [comment]
@vasslitvinov - would you be willing to also check the module code changes in ChapelReduce, ChapelSyncvar, ReplicatedVar?

File: modules/standard/IO.chpl [hunk 5]
```diff
@@ -5692,7 +5692,8 @@ class _channel_regexp_info {
 }
 
 pragma "no doc"
-proc channel._match_regexp_if_needed(cur:size_t, len:size_t, ref error:syserr, ref style:iostyle, ref r:_channel_regexp_info)
+proc channel._match_regexp_if_needed(cur:size_t, len:size_t, ref error:syserr,
```
Comment (mppf) [inline modules/standard/IO.chpl]: @psahabu - could you also review these I/O changes? Thanks.

File: modules/packages/TOML.chpl [hunk 6]
```diff
@@ -53,9 +53,9 @@ array. */
 proc parseToml(input: string) : Toml {
   var D: domain(string);
   var table: [D] Toml;
-  var rootTable = new Toml(table);
-  const source = new Source(input);
-  const parser = new Parser(source, rootTable);
+  var rootTable = new unmanaged Toml(table);
```
Comment (mppf) [inline modules/packages/TOML.chpl]: @ben-albrecht - could you review these changes to the TOML module? thanks

**Comment by vasslitvinov** [comment]
@mppf the changes in ChapelReduce, ChapelSyncvar, ReplicatedVar look trivial to me, so are OK.

**Comment by mppf** [comment]
@vasslitvinov - could you have a quick look at commit 3ae617d030e3a6cbf325a357b7608728a68f414c ? It's a late-breaking compiler change to resolve an issue with GASNet testing.

**Comment by vasslitvinov** [comment]
@mppf - 3ae617d looks trivial and OK to me.

**Pull request merged**